src/lsp/commands/vscode-commands.ts

```diff
@@ -0,0 +1,353 @@
+    return;
+  }
+
+  // Find the furthest valid root, excluding workspace folders (unless they are also valid roots).
+  const valid_project_roots = await project_utils
+    .findProjectRootsWithReasons({
+      include_workspace_folders: false,
+    })
+    .then((roots) => {
+      return filterOutRootsWithClients(roots, clients).map((root) => root.uri);
+    });
+  const selected = project_utils.findFurthestParent(document.uri, valid_project_roots);
+
+  const roots = await project_utils.findProjectRootsWithReasons().then((roots) => {
+    return filterOutRootsWithClients(roots, clients);
+  });
+
+  const selected_root = await project_utils.pickProjectRoot(roots, selected);
+  if (!selected_root) {
+    return;
+  }
+
+  await handler(selected_root);
+};
+
+const pickClient = async (clients: defs.LSPClientMap) => {
+  if (clients.size === 0) {
+    return;
+  }
+  if (clients.size === 1) {
+    return Array.from(clients.keys())[0];
+  }
+  const choices = Array.from(clients.keys()).map((uri) => {
+    return {
+      label: uri,
+    };
+  });
+  const selected_client = await vscode.window.showQuickPick(choices, { title: 'clojure-lsp' });
+  return selected_client?.label;
+};
+
+const stopClient = (clients: defs.LSPClientMap, id: string) => {
+  const client = clients.get(id);
+  return client?.stop().catch((err) => {
+    console.error(`Failed to stop client ${id}`, err);
+  });
+};
+
+const stopHandler = async (clients: defs.LSPClientMap, uri?: string) => {
+  if (uri) {
+    void stopClient(clients, uri);
+    return;
+  }
+  const id = await pickClient(clients);
+  if (!id) {
+    return;
+  }
+  void stopClient(clients, id);
+};
+
+const restartHandler = async (
+  clients: defs.LSPClientMap,
+  startHandler: (uri: vscode.Uri) => Promise<void>,
+  uri?: string
+) => {
+  const id = uri ? uri : await pickClient(clients);
+  if (!id) {
+    return;
+  }
+
+  await stopClient(clients, id);
+  clients.delete(id);
+
+  await startHandler(vscode.Uri.parse(id));
+};
+
+async function showServerInfo(clients: defs.LSPClientMap, id: string) {
+  const client = clients.get(id);
+
+  const serverInfo = await api.getServerInfo(client);
+  const calvaSaysChannel = state.outputChannel();
+  calvaSaysChannel.appendLine(`Clojure-lsp server info:`);
+  const serverInfoPretty = JSON.stringify(serverInfo, null, 2);
+  calvaSaysChannel.appendLine(serverInfoPretty);
+  calvaSaysChannel.show(true);
+}
+
+async function openLogFile(clients: defs.LSPClientMap, id: string) {
+  const client = clients.get(id);
+
+  const serverInfo = await api.getServerInfo(client);
+  const logPath = serverInfo['log-path'];
+  void vscode.window.showTextDocument(vscode.Uri.file(logPath));
+}
+
+const downloadServerHandler = async (context: vscode.ExtensionContext) => {
+  void vscode.window.showInformationMessage(`Downloading clojure-lsp server binary...`);
+  const version = await downloader.ensureLSPServer(context, true);
```

Comment:
```suggestion
  const path = await downloader.ensureLSPServer(context, true);
```
This function returns the clojure-lsp path.

Reply:
Resolved in 3f29c342